fish-rust/src/curses.rs

```diff
@@ -0,0 +1,261 @@
+
+        /// Checks for the presence of a termcap flag identified by the first two characters of
+        /// `id`. The C function returns an integer, but we just reinterpret that as a bool.
+        pub fn tgetflag(id: *const libc::c_char) -> bool;
+
+        /// Checks for the presence and value of a number capability in the termcap/termconf
+        /// database. A return value of `-1` indicates not found.
+        pub fn tgetnum(id: *const libc::c_char) -> libc::c_int;
+
+        pub fn tgetstr(
+            id: *const libc::c_char,
+            area: *mut *mut libc::c_char,
+        ) -> *const libc::c_char;
+    }
+}
+
+// String capabilities
+pub const ENTER_ITALICS_MODE: StringCap = StringCap::new("ZH");
+pub const EXIT_ITALICS_MODE: StringCap = StringCap::new("ZR");
+pub const ENTER_DIM_MODE: StringCap = StringCap::new("mh");
+
+// Number capabilities
+pub const MAX_COLORS: Number = Number::new("Co");
+
+// Flag capabilities
+pub const EAT_NEWLINE_GLITCH: Flag = Flag::new("xn");
+
+pub struct Term {
+    has_overrides: AtomicBool,
+    overrides: Mutex<Vec<(StringCap, Arc<CString>)>>,
+}
+
+impl Term {
+    /// Internal constructor function. Like `Default` but only usable from within the module.
+    fn new() -> Self {
+        Term {
+            has_overrides: AtomicBool::new(false),
+            overrides: Mutex::new(Vec::new()),
+        }
+    }
+
+    /// Looks up support for [`Capability`] `capability` in the termcap/terminfo database via the
+    /// curses library.
+    pub fn get<'a, C: Capability<'a>>(&'a self, capability: C) -> C::Result {
+        capability.lookup(self)
+    }
+
+    /// Overrides the string value of `capability` for the current terminal.
+    pub fn set(&self, id: StringCap, value: &str) {
+        let value = CString::new(value).unwrap().into();
+        let mut overrides = self.overrides.lock().expect("Mutex poisoned!");
+        match overrides.binary_search_by(|entry| entry.0.cmp(&id)) {
+            Ok(idx) => overrides[idx] = (id, value),
+            Err(idx) => overrides.insert(idx, (id, value)),
+        }
+        self.has_overrides.store(true, Ordering::Relaxed);
+    }
+}
+
+pub trait Capability<'a> {
+    type Result: Sized + 'a;
+    fn lookup(&self, term: &'a Term) -> Self::Result;
+}
+
+impl StringCap {
+    fn sys_lookup<'a>(&self) -> Option<ArcCow<'a, CStr, CString>> {
+        let result = unsafe {
+            const NULL: *const i8 = core::ptr::null();
+            match sys::tgetstr(self.0.as_ptr(), core::ptr::null_mut()) {
+                NULL => return None,
+                // termcap spec says nul is not allowed in terminal sequences and must be encoded,
+                // so we can safely unwrap here.
+                result => CStr::from_ptr(result),
+            }
+        };
+        Some(ArcCow::Borrowed(result))
+    }
+}
+
+impl<'a> Capability<'a> for StringCap {
+    type Result = Option<ArcCow<'a, CStr, CString>>;
```

Comment:
Actually, since I changed from `&str` to `Arc<CString>` for the reasons explained above, I can drop the mutex and change the vec to `Vec<Cell<Arc<CString>>>` - everything stays the same but we don't need a mutex since the clone out of the cell is "free". I can also go back to caching the result of ncurses lookups since there would never be a mutex so we don't need to avoid caching in order to avoid locking, meaning we always return `Arc<CString>` so there's no need for the `Cow`.